### PR TITLE
Add caching to lychee links checker

### PR DESCRIPTION
**Pull request by dmathieu**
This adds caching to the links checker, to hopefully reduce the false positive failures due to 429s.

### CODE REVIEW

**Comment by dmathieu** [comment]
I'm currently basing the cache on the GIT commit, but we can make it global if needed.

File: .github/workflows/links-fail-fast.yml
Comment (pellared) [inline .github/workflows/links-fail-fast.yml:25]: After 1 day the cache would always return an out-dated `.lycheecache`.
See: https://github.com/open-telemetry/opentelemetry-go/actions/runs/8596315520/job/23552862599?pr=5160

I think we should "force-update" the cache. I see that there is a `save-always: true` option, but I am not sure if it does what we want. Maybe we would have to use `actions/cache/save@v3` ([docs](https://github.com/actions/cache/blob/main/save/README.md)).

**Review by pellared — CHANGES_REQUESTED** [review]

File: .github/workflows/links-fail-fast.yml
Comment (dmathieu) [reply .github/workflows/links-fail-fast.yml:15]: Why? If some checks succeed and one of them fails (due to a 429), caching the succeeding ones means the next run has more chances of succeeding.

File: .github/workflows/links-fail-fast.yml
Comment (pellared) [reply .github/workflows/links-fail-fast.yml:24]: I think that in your run the key was different. I guess it was run then the key was generated from the SHA so it was always storing a new cache.

> Cache restored from key: cache-lychee-56719d07e5252211b258aa302f4a7e53da44f8de

Comment (pellared) [reply .github/workflows/links-fail-fast.yml:24]: We can rerun the action from this PR in one day and see what happens 😉

**Comment by dmathieu** [comment]
> Won't caching on the git sha mean a new cache is used for each commit? That will make the cache ineffective as it will be new for each commit, right?

Based on the lychee and actions/cache documentations, I don't think that's accurate.
https://github.com/lycheeverse/lychee-action?tab=readme-ov-file#utilising-the-cache-feature

We do store the sha in the cache name, but if there is no cache hit, the `restore-keys` argument means we'll still restore the most recent one, no matter the sha.
See https://github.com/actions/cache?tab=readme-ov-file#inputs

**Comment by dmathieu** [comment]
Discussing with @pellared, we have decided to only cache if the run is fully successful, because lychee caches everything, including failures.

File: .github/workflows/links.yml
Comment (pellared) [inline .github/workflows/links.yml:29]:
```suggestion
          args: --max-concurrency 1 --cache --max-cache-age 1d .
```

How about this to reduce the possibility of rate limiting? 

This job can run longer as it is a scheduled (or manually triggered) one.

**Review by pellared — APPROVED** [review]

**Comment by MrAlias** [comment]
> > Won't caching on the git sha mean a new cache is used for each commit? That will make the cache ineffective as it will be new for each commit, right?
> 
> Based on the lychee and actions/cache documentations, I don't think that's accurate. https://github.com/lycheeverse/lychee-action?tab=readme-ov-file#utilising-the-cache-feature
> 
> We do store the sha in the cache name, but if there is no cache hit, the `restore-keys` argument means we'll still restore the most recent one, no matter the sha. See https://github.com/actions/cache?tab=readme-ov-file#inputs

Doesn't this mean that we will always use the last cache from "restore-keys" in the normal operation? We almost never re-run the link checker on the same commit.

If this is the case using a global cache would be equivalent and reduce the number of caches stored in actions.

**Comment by pellared** [comment]
> If this is the case using a global cache would be equivalent and reduce the number of caches stored in actions.

Can you please describe what do you mean by "a global cache"?

Do you mean using a single cache key value? The problem with it is that the action does not save an updates the cache when the cache was restored using the key. See https://github.com/open-telemetry/opentelemetry-go/pull/5160#discussion_r1555408664

**Comment by MrAlias** [comment]
> Do you mean using a single cache key value? The problem with it is that the action does not save an updates the cache when the cache was restored using the key. See [#5160 (comment)](https://github.com/open-telemetry/opentelemetry-go/pull/5160#discussion_r1555408664)

:+1:

**Review by MrAlias — APPROVED** [review]

**Comment by pellared** [comment]
Waiting for an non-Splunker approval.

**Review by dashpole — APPROVED** [review]
I haven't looked too deeply into lychee, but approving since this follows https://github.com/lycheeverse/lychee-action?tab=readme-ov-file#utilising-the-cache-feature, and based on other approvals.

**Review by XSAM — APPROVED** [review]

**Pull request merged**
